lib/src/fixed_datetime_codec.dart

```diff
@@ -0,0 +1,153 @@
+/// specification in https://www.unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table,
+/// the format string (or "skeleton") `YYY` specifies only the padding, so
+/// `1996` would be a valid match. This limits is use to format strings
+/// containing delimiters, as the parser would not know how many digits to take
+/// otherwise.
+///
+/// Also, this parser does not know about locales and parses to the current
+/// locale only.
+class FixedDateTimeCodec {
+  static const _validChars = ['Y', 'M', 'D', 'E', 'C', 'h', 'm', 's'];
+
+  final String _pattern;
+  final _occurences = <String, _Range>{};
+
+  FixedDateTimeCodec(this._pattern) {
+    String current = '';
+    for (var i = 0; i < _pattern.length; i++) {
+      var char = _pattern[i];
+      if (!_validChars.contains(char)) {
+        current = '';
+        continue;
+      }
+      var newChar = current != char;
+      if (newChar) {
+        var hasNotSeenBefore = !_occurences.containsKey(char);
+        if (hasNotSeenBefore) {
+          _occurences[char] = _Range(i, i);
+        } else {
+          throw Exception(
+              'The _pattern string "$_pattern" contains multiple instances of the formatting char block $char, both at position ${_occurences[char]!.from} and $i');
+        }
+      }
+      current = char;
+      _occurences.update(current, (value) => _Range(value.from, value.to + 1));
+    }
+  }
+
+  /// Convert a datetime to a string exactly as specified by the [_pattern].
+  String encode(DateTime dt) {
+    var startingPoints =
+        _occurences.map((key, value) => MapEntry(value.from, key));
+    var sb = StringBuffer();
+    for (var i = 0; i < _pattern.length; i++) {
+      if (startingPoints.containsKey(i)) {
+        var key = startingPoints[i];
+        var length = _occurences[key]!.length;
+        var number = _extractNumFromDateTime(key, dt);
+        var numAsString = number.toString();
+        if (numAsString.length > length) {
+          throw Exception(
+              "The datetime $dt cannot be parsed as $number is longer than the _pattern $key allows");
+        } else {
+          sb.write(numAsString.padLeft(length, '0'));
+        }
+        i += length - 1;
+      } else {
+        sb.write(_pattern[i]);
+      }
+    }
+    return sb.toString();
+  }
+
+  int _extractNumFromDateTime(String? key, DateTime dt) {
+    switch (key) {
+      case 'Y':
+        return dt.year;
+      case 'C':
+        return (dt.year / 100).floor();
+      case 'E':
+        return (dt.year / 10).floor();
+      case 'M':
+        return dt.month;
+      case 'D':
+        return dt.day;
+      case 'h':
+        return dt.hour;
+      case 'm':
+        return dt.minute;
+      case 's':
+        return dt.second;
+    }
+    throw ArgumentError.value(key);
+  }
+
+  /// Parse a string [dateTimeStr] to a local [DateTime] as specified in the
+  /// [_pattern]. Throws an exception on failure.
+  DateTime decodeToLocal(String dateTimeStr) {
```

Comment:
I am also not sure about how to name the methods and the class, I went with the Codec/Decode/Encode, ??/format/parse might make more sense, I don't know.

Reply:
I'd use `decode` here. The `toLocal` is not what `Codec` does. I'll make the choice between local time and UTC time in the constructor.

I'd then also have a `tryDecode` below. Which is a little weird because it's *normally* called `tryParse`, but it's internally consistent.